### ISSUE TITLE
Per channel quantization for the weights

### ISSUE THREAD

**Issue by Nimitz14** [pytorch/QNNPACK#44]
Hiya,

Is it possible to provide quantization parameters on a finer scale than per layer? In for example (1) they talk about doing per channel quantization for the weights, they're talking about convolutions there, but I remember reading a paper where they mentioned doing this for a normal linear transform as well (can't find the paper again at the moment sorry). Any plans for including such capabilities?

I'm trying to do post training quantization but am getting significant degradation when replacing a single layer. It's a smaller network and from what I've read these are harder to quantise. If you don't mind me asking I'm curious what your experience is with post-training quantization and how well it works?

(1) https://arxiv.org/pdf/1806.08342.pdf

**Comment by Maratyszcza** [comment]
We considered supporting per-channel quantization, but haven't made a decision yet. It requires non-trivial changes to QNNPACK, and accuracy loss is already very small if the model is trained with quantization-aware training. Even if we do support per-channel quantization, it won't be soon, and we'll have to implement a full set of ops for symmetric quantization to make it usable.

I suggest you try to train your model with quantization-aware training, as described in [this paper](https://arxiv.org/abs/1712.05877) by Benoit Jacob et al.
